Move to TOML for manifest and lock

carolynvs opened the pull request:
Fixes #119; fixes #168.

I'm doing this in stages so that we can identify when I'm off in the weeds or need to make more decisions about the config file format.

- [x] Vendor https://github.com/pelletier/go-toml.
- [x] Write manifest and lock in TOML format.
- [x] Read manifest and lock in TOML format (all tests but diff should pass)
- [x] Remove old JSON files and logic.
- [x] Fix output of dry-run to use TOML (diff test and all others should pass)
- [x] (Optional) Bike-shed the look of the generated TOML, e.g. inline tables, indenting, key sorting, etc.

carolynvs commented:
@sdboyer @tro3  I didn't find any reflection-based marshal magic for converting between structs and TOML in this go-toml package. If we want to go with this lib, it may make sense to contribute that functionality back into the library. For the moment, I am explicitly performing the mapping in `MarshalTOML`.

https://github.com/golang/dep/pull/342/files#diff-8e0a04a8bff74b300db5882b4be50589R133

Do you think that makes sense or should I try out another lib?

sdboyer commented:
@carolynvs hmm...that puts me on the fence about it, as I'd generally rather have the (presumably cross-TOML implementation portable) struct tags as the canonical way of expressing this information, to minimize effort in the event that we switch TOML libs later. idk how much work that's really saving, though, so it seems like a soft requirement at worst.

Maybe some input from @pelletier here would be good - any plans to add reflection-based struct tag support to go-toml?

sdboyer commented:
Eh, that was too wishy-washy, sorry. If we can have struct tags and reflection-based parsing, great. If not, well, it seems like the one-off implementation here isn't too awful. Either way, not a reason to derail the current track.

Is there are reason you're keeping the JSON struct tags around?

pelletier commented:
@sdboyer yes, reflection-based (un)marshaling is the next thing that I'd like to implement in go-toml. not eta on that though, but pull requests are welcome :)

tro3 commented:
@pelletier - I have some time coming up the next few weeks.  Mind if I take a crack at the reflection-based marshaling?  If so, I'll open an issue at go-toml and get you a WIP PR in the short term to make sure you approve of the structure before I dive in too deep.  Thoughts?

pelletier commented:
@tro3 sounds great!

carolynvs commented:
@sdboyer No reason, just hadn't gotten to removing all the json stuff yet. 😃  I've added a task to the PR to make that more clear.

carolynvs commented:
* Changed manifest format. `dependencies` and `overrides` use an array of tables instead of maps now, which is easier to work with.

    ```toml
    ignores = ["github.com/foo/bar"]
    
    [[dependencies]]
      name = "github.com/sdboyer/gps"
      version = ">=0.12.0, <1.0.0"
    
    [[dependencies]]
      name = "github.com/babble/brook"
      revision = "d05d5aca9f895d19e9265839bffeadd74a2d2ecb"
    
    [[overrides]]
      branch = "master"
      name = "github.com/sdboyer/gps"
      source = "https://github.com/sdboyer/gps"
    ````
* Update all the test files from JSON to TOML
* Poor man's mapping from TOML to structs. Can't wait for https://github.com/pelletier/go-toml/pull/149! ❤️

omeid commented:
Couldn't now overrides be just an attribute on the dependency itself?

carolynvs commented:
@omeid Let's not make changes to the config file structure in this PR, beyond the initial scope of switching from JSON to TOML. 😀 It's already a pretty big change, and I want to limit the amount of regressions/confusion it may cause.

nathany commented:
For generated TOML, I think it would be nice to use inline tables to get each dependency onto a single line.

I'm think that may work better for diffs/merges, though I'm not totally sure. Also it's nice for rearranging the dependencies by movings lines up/down in an editor (for super organized types).

carolynvs commented:
@nathany I think that depends on how we feel about long lines in the file. Personally, I find it hard to read.

For example, here's what I think the lock would look like with inline tables. Note, I am cheating a bit by using `"."` for the package list, to make the shortest possible entry.

```
memo = ""
projects = [
 { name = "github.com/sdboyer/deptest", packages = ["."], revision = "ff2948a2ac8f538c4ecd55962e919d1e13e74baf", version = "v0.8.0" }
 { name = "github.com/sdboyer/deptestdos", packages = ["."], revision = "a0196baa11ea047dd65037287451d36b861b00ea", version = "1.0.0" }
]
```

Though I'm not 100% sure that's how an array of inline tables should look? Either way support for changing the formatter in go-toml would be needed, not sure about the parser.

nathany commented:
Okay. Those are some pretty long lines.

It's probably best to just use it for a while to see how well diffs/merges work with the multi-line syntax. Thanks

carolynvs commented:
I've updated the lock diff to output matching TOML. I've also tried out @tro3 's reflection based mapping, it's super close and I think it's worth waiting for it, removes a TON of code. 😃

pelletier commented:
https://github.com/pelletier/go-toml/pull/149 has been merged, bringing reflection-based marshaling goodness to go-toml. props to @tro3 :)

sdboyer commented:
We need to roll the file name change into this, too, so that we minimize the amount of change that goes in at once. We're still pondering those, but...

Cursory look has me thinking this is pretty good. I'll look more soon, I hope later tonight :)

sdboyer commented on analyzer_test.go:
```diff
@@ -18,7 +18,7 @@ func TestDeriveManifestAndLock(t *testing.T) {
 
 	h.TempDir("dep")
-	golden := "analyzer/manifest.json"
+	golden := "analyzer/manifest.toml"
```
I realize you didn't introduce this, but now that I'm seeing it, I believe it should be `filepath.Join("analyzer", ManifestName)`.

tro3 replied:
FWIW, being a PC guy, I've found Go does a startingly good job of using '/' in a platform-independent manner  in path strings.  But `filepath.Join` is still likely more robust.

sdboyer commented on cmd/dep/testdata/harness_tests/ensure/empty/case1/final/lock.toml:
```diff
@@ -0,0 +1,7 @@
+memo = "63510efb9632ec69c1164ce396d7ebea4ad3884b4fa508373da17226d5a39739"
+
+[[projects]]
+  name = "github.com/sdboyer/deptest"
```
Hmm...yeah, I hate to bikeshed, but these indents are kinda making me twitch. What's your thinking behind indenting vs. not? Are there any typical standards out there for it? Do we know anything about, say, widespread editor defaults for indentation in TOML files?

tro3 replied:
The go-toml package will indent on `-update`.  (It will read fine with or without indent.)  Don't know if there is a setting to stop it - never looked.

carolynvs replied:
Yeah, preferably `go-toml` would let us specify formatting options. I'll double-check that doesn't exist already and see if it's an easy thing to add.

carolynvs replied:
Okay, it's not exposed but seems like it wouldn't be hard to add a flag to control the auto-indentation that it's currently doing. I'll submit a PR to `go-toml` for that.

sdboyer commented on cmd/dep/testdata/harness_tests/init/case1/final/lock.toml:
```diff
@@ -0,0 +1,12 @@
+memo = ""
```
Any idea how this one ended up without a memo? Mostly just curious.

carolynvs replied:
I am not sure why the memo is empty? 😄  It's been like that for a while; here's there current `lock.json` for that test:

https://github.com/golang/dep/blob/master/cmd/dep/testdata/harness_tests/init/case1/final/lock.json#L2

Think that's a bug?

sdboyer commented on fs.go:
```diff
@@ -16,6 +15,13 @@ import (
+// Marshaler is the interface implemented by types that
+// can marshal themselves into valid TOML.
+// TODO(carolynvs) Consider adding this to go-toml?
+type Marshaler interface {
```
Probably better not to export this as an interface, unless we have to. As your TODO notes, if it's going to be exported, that should be done by go-toml. If we can get away with exporting, `marshaler`, or better `tomlMarshaler` would be preferable.

carolynvs replied:
Thanks for noticing this! I had intended to put it in `go-toml` but that slipped my mind... 😊

carolynvs replied:
I ended up needing this to use reflection for the `LockDiff`. Here's the go-toml PR: https://github.com/pelletier/go-toml/pull/151

sdboyer commented on manifest.go:
```diff
@@ -23,143 +24,170 @@ type Manifest struct {
-	for n, pp := range rm.Dependencies {
-		m.Dependencies[gps.ProjectRoot(n)], err = toProps(n, pp)
+	for i := 0; i < len(raw.Dependencies); i++ {
+		name, prj, err := toProject(raw.Dependencies[i])
```
The cost of moving to an array from a map is that it's now possible for users to express the same project more than once. This loop has to validate that each entry in the input is unique, and fail hard otherwise.

sdboyer commented on manifest.go:
```diff
@@ -23,143 +24,170 @@ type Manifest struct {
 
-	for n, pp := range m.Ovr {
-		raw.Overrides[string(n)] = toPossible(pp)
+// TODO(carolynvs) when gps is moved, we can use the unexported gps.sortedConstraints
```
It'll still be unexported within the gps package, so probably not 😄

sdboyer commented on test/test.go:
```diff
@@ -584,7 +584,7 @@ func (h *Helper) Cleanup() {
 // ReadManifest returns the manifest in the current directory.
 func (h *Helper) ReadManifest() string {
-	m := filepath.Join(h.pwd(), "manifest.json")
+	m := filepath.Join(h.pwd(), "manifest.toml")
```
Use the constant

carolynvs replied:
I don't believe I can use the constant here (in the test pkg) because that would create a circular dependency?

sdboyer replied:
ahhh, totally, yes. my bad :)

sdboyer replied:
let's create a local constant in this test package too, and use that. it's not as good as one constant, but it's still better than string literals all over.

sdboyer commented on test/test.go:
```diff
@@ -594,7 +594,7 @@ func (h *Helper) ReadManifest() string {
 // ReadLock returns the lock in the current directory.
 func (h *Helper) ReadLock() string {
-	l := filepath.Join(h.pwd(), "lock.json")
+	l := filepath.Join(h.pwd(), "lock.toml")
```
Use the constant

sdboyer commented on testdata/txn_writer/expected_diff_output.txt:
```diff
@@ -1,34 +1,31 @@
-    }
-]
+
+Add:
```
Hot diggity damn these look so much better 😄

sdboyer commented on txn_writer.go:
```diff
@@ -113,12 +113,52 @@ func (diff *LockDiff) Format() (string, error) {
-	Branch   *StringDiff     `json:"branch,omitempty"`
-	Revision *StringDiff     `json:"revision,omitempty"`
-	Packages []StringDiff    `json:"packages,omitempty"`
+	Name     gps.ProjectRoot
```
these don't need toml struct tags?

carolynvs replied:
Sorry an emergency came up and I stopped mid review. I'll start back on this today. The current toml mapper doesn't yet support allowing a nested type to use a custom marshal function, so it was manual. I started adding that this weekend and hopefully can submit that to go-toml soon and use it.

carolynvs commented:
@sdboyer You mentioned rolling up the [filename changes](https://github.com/golang/dep/issues/168) into this PR to limit churn. Do we have final names yet?

carolynvs commented:
@sdboyer @tro3 Right now `go-toml` doesn't preserve the ordering of the fields when marshaling, instead simply outputting them in alphabetical order:

```toml
[[projects]]
  branch = "master"
  name = "github.com/pelletier/go-buffruneio"
  packages = ["."]
  revision = "c37440a7cf42ac63b919c752ca73a85067e05992"
```

instead of this (which matches the ordering of the fields as defined on the struct)

```toml
[[projects]]
  name = "github.com/pelletier/go-buffruneio"
  branch = "master"
  revision = "c37440a7cf42ac63b919c752ca73a85067e05992"
  packages = ["."]
```

Is this something that I need to address as part of this PR, in the name of not changing the file structure again later? It may be a big change to how the `TomlTree` is represented internally and could be a lot of work, so I wanted to get your thoughts (or maybe @tro3 you see a clean way of supporting that)?

carolynvs commented:
* Rebased, resolving the merge conflicts from updating `gps`
* Added constants for the file names to the lock package
* Updated LockDiff formatting to the toml struct tags

Looking into making the auto-indentation configurable next.

tro3 commented:
Is it really needed?  Don't know if he key order is important. 

In any case, I think it's an easy tweak to go-toml, as I recall an explicit sorting of the keys. If correct, this would just have to be bypassed with a configure option of some sort. 

I'm on vacation though, so if it needs to be me, it will have to wait a while.

carolynvs commented:
I spoke with @sdboyer and we will leave off trying to make the toml file format "pretty", so I am not going to try to tackle the indenting or sorting in this PR. I believe I've hit all the rest of the review feedback so it's ready for another look.

sdboyer commented:
OK for filenames, let's do `Gopkg.toml` and `Gopkg.lock`.

carolynvs commented:
File renaming is in! 💥

bradleyfalzon commented:
Is it possible, or was it already decided not, to read the old format and update the relevant files?

As the migration path, as I understand it, would be to remove the old files and generate new lock files - which would lock to different revisions?

Detecting existing files, reading them and rewriting to new files and new format would obviously be ideal from an early user's perspective. Perhaps a flag, or a separate application (depfix)?

I don't see this as a major requirement at all, just asking.

sdboyer commented:
Sorry for delay on merge, here. I was trying to get things together with a blog for dep (re: #294), and I put this in the queue behind doing that. Looks like I should have that blog done tonight, and I plan to merge this once it's all set up 😄

sdboyer commented:
And in, finally. Thanks! 🎉 🎉 🎉 🎉
